File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/MetricsRouteProgress.java
package com.mapbox.services.android.navigation.v5.routeprogress;

import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.mapbox.api.directions.v5.models.DirectionsRoute;
import com.mapbox.api.directions.v5.models.LegStep;
import com.mapbox.api.directions.v5.models.RouteLeg;
import com.mapbox.api.directions.v5.models.StepManeuver;
import com.mapbox.geojson.Point;

public class MetricsRouteProgress {

  private int directionsRouteDistance;
  private int directionsRouteDuration;
  private String directionsRouteProfile;
  private Point directionsRouteDestination;
  private int distanceRemaining;
  private int durationRemaining;
  private int distanceTraveled;
  private int currentStepDistance;
  private int currentStepDuration;
  private int currentStepDistanceRemaining;
  private int currentStepDurationRemaining;
  private String currentStepName;

  private String upcomingStepInstruction;
  private String upcomingStepModifier;
  private String upcomingStepType;
  private String upcomingStepName;

  private String previousStepInstruction;
  private String previousStepModifier;
  private String previousStepType;
  private String previousStepName;

  public MetricsRouteProgress(@Nullable RouteProgress routeProgress) {
    if (routeProgress != null) {
      obtainRouteData(routeProgress.directionsRoute());
      obtainLegData(routeProgress.currentLegProgress());
      obtainStepData(routeProgress);
      this.distanceRemaining = (int) routeProgress.distanceRemaining();
      this.durationRemaining = (int) routeProgress.durationRemaining();
      this.distanceTraveled = (int) routeProgress.distanceTraveled();
    } else {
      initDefaultValues();
    }
  }

  private void initDefaultValues() {
    directionsRouteProfile = "";
    directionsRouteDestination = Point.fromLngLat(0d, 0d);
    currentStepName = "";
    upcomingStepInstruction = "";
    upcomingStepModifier = "";
    upcomingStepType = "";
    upcomingStepName = "";
    previousStepInstruction = "";
    previousStepModifier = "";
    previousStepType = "";
    previousStepName = "";
  }

  private void obtainRouteData(DirectionsRoute route) {
    directionsRouteDistance = route.distance() != null ? route.distance().intValue() : 0;
    directionsRouteDuration = route.duration() != null ? route.duration().intValue() : 0;
    directionsRouteProfile = hasRouteProfile(route) ? route.routeOptions().profile() : "";
    directionsRouteDestination = retrieveRouteDestination(route);
  }

  private void obtainLegData(RouteLegProgress legProgress) {
    currentStepDistance = (int) legProgress.currentStep().distance();
    currentStepDuration = (int) legProgress.currentStep().duration();
    currentStepDistanceRemaining = (int) legProgress.currentStepProgress().distanceRemaining();
    currentStepDurationRemaining = (int) legProgress.currentStepProgress().durationRemaining();
    currentStepName = hasStepName(legProgress) ? legProgress.currentStep().name() : "";
  }

  private void obtainStepData(RouteProgress routeProgress) {
    RouteLegProgress legProgress = routeProgress.currentLegProgress();
    if (legProgress.upComingStep() != null) {
      upcomingStepName = legProgress.upComingStep().name();
      StepManeuver upcomingManeuver = legProgress.upComingStep().maneuver();
      if (upcomingManeuver != null) {
        upcomingStepInstruction = upcomingManeuver.instruction();
        upcomingStepType = upcomingManeuver.type();
        upcomingStepModifier = upcomingManeuver.modifier();
      }
    }
    StepManeuver currentManeuver = legProgress.currentStep().maneuver();
    if (currentManeuver != null) {
      previousStepInstruction = currentManeuver.instruction();
      previousStepType = currentManeuver.type();
      previousStepModifier = currentManeuver.modifier();
    }
    previousStepName = currentStepName;
  }

  private boolean hasRouteProfile(DirectionsRoute route) {
    return route.routeOptions() != null && !TextUtils.isEmpty(route.routeOptions().profile());
  }

  private Point retrieveRouteDestination(DirectionsRoute route) {
    RouteLeg lastLeg = route.legs().get(route.legs().size() - 1);
    LegStep lastStep = lastLeg.steps().get(lastLeg.steps().size() - 1);
    StepManeuver finalManuever = lastStep.maneuver();
    if (finalManuever.location() != null) {
      return finalManuever.location();
    }
    return Point.fromLngLat(0d, 0d);
  }

  private boolean hasStepName(RouteLegProgress legProgress) {
    return legProgress.currentStep().name() != null && !TextUtils.isEmpty(legProgress.currentStep().name());
  }

  public int getDirectionsRouteDistance() {
    return directionsRouteDistance;
  }

  public int getDirectionsRouteDuration() {
    return directionsRouteDuration;
  }

  public String getDirectionsRouteProfile() {
    return directionsRouteProfile;
  }

  public Point getDirectionsRouteDestination() {
    return directionsRouteDestination;
  }

  public int getDistanceRemaining() {
    return distanceRemaining;
  }

  public int getDurationRemaining() {
    return durationRemaining;
  }

  public int getDistanceTraveled() {
    return distanceTraveled;
  }

  public int getCurrentStepDistance() {
    return currentStepDistance;
  }

  public int getCurrentStepDuration() {
    return currentStepDuration;
  }

  public int getCurrentStepDistanceRemaining() {
    return currentStepDistanceRemaining;
  }

  public int getCurrentStepDurationRemaining() {
    return currentStepDurationRemaining;
  }

  public String getUpcomingStepInstruction() {
    return upcomingStepInstruction;
  }

  public String getUpcomingStepModifier() {
    return upcomingStepModifier;
  }

  public String getUpcomingStepType() {
    return upcomingStepType;
  }

  public String getUpcomingStepName() {
    return upcomingStepName;
  }

  public String getPreviousStepInstruction() {
    return previousStepInstruction;
  }

  public String getPreviousStepModifier() {
    return previousStepModifier;
  }

  public String getPreviousStepType() {
    return previousStepType;
  }

  public String getPreviousStepName() {
    return previousStepName;
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/routeprogress/MetricsRouteProgress.kt
package com.mapbox.services.android.navigation.v5.internal.navigation.routeprogress

import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.geojson.Point
import com.mapbox.services.android.navigation.v5.routeprogress.RouteLegProgress
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
import com.mapbox.services.android.navigation.v5.utils.extensions.ifNonNull

class MetricsRouteProgress(routeProgress: RouteProgress?) {

    companion object {
        private val DEFAULT_POINT = Point.fromLngLat(0.0, 0.0)
    }

    var directionsRouteDistance: Int = 0
        private set
    var directionsRouteDuration: Int = 0
        private set
    var directionsRouteProfile: String = ""
        private set
    var directionsRouteDestination: Point = DEFAULT_POINT
        private set
    var distanceRemaining: Int = 0
        private set
    var durationRemaining: Int = 0
        private set
    var distanceTraveled: Int = 0
        private set

    var currentStepDistance: Int = 0
        private set
    var currentStepDuration: Int = 0
        private set
    var currentStepDistanceRemaining: Int = 0
        private set
    var currentStepDurationRemaining: Int = 0
        private set
    private var currentStepName: String = ""

    var upcomingStepInstruction: String? = null
        private set
    var upcomingStepModifier: String? = null
        private set
    var upcomingStepType: String? = null
        private set
    var upcomingStepName: String? = null
        private set

    var previousStepInstruction: String? = null
        private set
    var previousStepModifier: String? = null
        private set
    var previousStepType: String? = null
        private set
    var previousStepName: String = ""
        private set

    var legIndex: Int = 0
        private set
    var legCount: Int = 0
        private set
    var stepIndex: Int = 0
        private set
    var stepCount: Int = 0
        private set

    init {
        ifNonNull(
            routeProgress, routeProgress?.directionsRoute(),
            routeProgress?.currentLegProgress(),
            routeProgress?.distanceRemaining(), routeProgress?.durationRemaining()
        ) { _routeProgress, _directionsRoute, _currentLegProgress, _distanceRemaining, _durationRemaining ->
            obtainRouteData(_directionsRoute)
            obtainLegData(_currentLegProgress)
            obtainStepData(_routeProgress)
            distanceRemaining = _distanceRemaining.toInt()
            durationRemaining = _durationRemaining.toInt()
            distanceTraveled = _routeProgress.distanceTraveled()?.toInt() ?: 0
            legIndex = _routeProgress.legIndex() ?: 0
            legCount = _directionsRoute.legs()?.size ?: 0
            stepIndex = _currentLegProgress.stepIndex() ?: 0
            stepCount = _routeProgress.currentLeg()?.steps()?.size ?: 0
        } ?: initDefaultValues()
    }

    private fun initDefaultValues() {
        directionsRouteProfile = ""
        directionsRouteDestination = DEFAULT_POINT
        currentStepName = ""
        upcomingStepInstruction = ""
        upcomingStepModifier = ""
        upcomingStepType = ""
        upcomingStepName = ""
        previousStepInstruction = ""
        previousStepModifier = ""
        previousStepType = ""
        previousStepName = ""
    }

    private fun obtainRouteData(route: DirectionsRoute) {
        directionsRouteDistance = route.distance()?.toInt() ?: 0
        directionsRouteDuration = route.duration()?.toInt() ?: 0
        directionsRouteProfile = route.routeOptions()?.profile() ?: ""
        directionsRouteDestination = retrieveRouteDestination(route)
    }

    private fun obtainLegData(legProgress: RouteLegProgress) {
        currentStepDistance = legProgress.currentStep()?.distance()?.toInt() ?: 0
        currentStepDuration = legProgress.currentStep()?.duration()?.toInt() ?: 0
        currentStepDistanceRemaining =
            legProgress.currentStepProgress()?.distanceRemaining()?.toInt() ?: 0
        currentStepDurationRemaining =
            legProgress.currentStepProgress()?.durationRemaining()?.toInt() ?: 0
        currentStepName = legProgress.currentStep()?.name() ?: ""
    }

    private fun obtainStepData(routeProgress: RouteProgress) {
        val legProgress = routeProgress.currentLegProgress()
        legProgress?.upComingStep()?.let { upcomingStep ->
            upcomingStepName = upcomingStep.name()
            upcomingStep.maneuver().let {
                upcomingStepInstruction = it.instruction()
                upcomingStepType = it.type()
                upcomingStepModifier = it.modifier()
            }
        }
        legProgress?.currentStep()?.maneuver().let { stepManeuver ->
            previousStepInstruction = stepManeuver?.instruction()
            previousStepType = stepManeuver?.type()
            previousStepModifier = stepManeuver?.modifier()
        }
        previousStepName = currentStepName
    }

    private fun retrieveRouteDestination(route: DirectionsRoute): Point =
        route.legs()?.lastOrNull()?.steps()?.lastOrNull()?.maneuver()?.location()
            ?: DEFAULT_POINT
}
